# Operadores e variáveis

Vamos começar a programar em Python escrevendo a parte de um programa que faz o cálculo dos pontos de um jogador em um jogo com temática espacial, com naves e robôs, estilo *"shoot them up!"*, daquele tipo que a gente atira em tudo que se mexer.

Depois vamos aprender a usar variáveis para melhorar essses trechos do programa. Variáveis são uma forma de guardarmos informações na memória do computador para usarmos depois. 

Além disso, elas também permitem que a gente use elas ao invés de usarmos números ou outros dados que um programa manipula, com a vantagem de que as variáveis podem mudar de valor (podem *variar*) e números e outros dados costumam ser fixos. 

Por exemplo, o valor do placar de um jogo não pode ser um valor fixo, porque o placar vai mudando de valor ao longo do jogo. Para isso precisamos usar uma variável ao invés de um valor para programar um placar.

Nos cálculos vamos usar as operações aritméticas básicas de adição `+`, subtração `-`, divisão `÷` e multiplicação `x`.

## Fazendo cálculos

Em Python fazemos cálculos quase do mesmo modo como fazemos fora do computador, com papel e caneta.

Veja no primeiro exemplo abaixo:

> Para rodar o código Python abaixo (Windows ou Linux), clique na célula que contém o código e tecle <kbd>Ctrl+Enter</kbd> (no Mac, <kbd>Cmd+Return</kbd>). Faça isso com todos os exemplos de código que vierem. Assim você acompanha e experimenta a matéria ao mesmo tempo. 

In [3]:
2 * 10

20

Essa linha de código calculou o *produto* de `2` e `10`. É uma multiplicação. Só repare que a operação de multiplicação é representada por um *asterisco* e não pelo `xis`, ou o `.` que estamos acostumados a ver.

## Operadores aritméticos do Python

Os operadores aritméticos do Python são:

- `+` - soma (adição)
- `-` - subtração 
- `/` - divisão (razão)
- `*` - multiplicação (produto)

## Calculando os pontos de uma jogada

Vamos imaginar que você é um mago caçador de robôs feiticeiros espaciais. Para cada robô que você captura você acumula 0.05 pontos de experiência e 0.5 pontos de magia. 

<center>

!['robô feiticeiro 1'](img/robof-64px.png)
!['robô feiticeiro 2'](img/robof-64px.png)
!['robô feiticeiro 3'](img/robof-64px.png)
!['robô feiticeiro 4'](img/robof-64px.png)
!['mago'](img/mago-64px.png)

</center>

Agora imagine que em uma jogada você capturou 4 robôs feiticeiros ao mesmo tempo. Quantos pontos de experiência e de magia você ganhou?


Vamos ver no próximo exemplo como calcular isso com o Python.

> As linhas que começam por um `#` não são executadas pelo Python. Elas são usadas para acrescentar comentários ao código.

In [4]:
# Você capturou 4 robôs feiticeiros
# e cada um deles vale 0.05 pontos de experiência
# então

4 * 0.05

# Rode o código para ver quantos pontos 
# de experiência você ganhou na jogada

0.2

Então, você ganhou `0.2` pontos de experiência na jogada.

Para calcular quantos pontos de magia você ganhou, multiplique o número de robôs que você capturou pelo número de pontos de magia que você ganha por cada um deles (que é `0.5`)

Na célula abaixo, escreva a operação que faz esse cálculo.

In [5]:
4 * 0.5

2.0

O resultado deve ser `2.0` pontos de magia. Confere com a resposta da sua operação?

## Um cálculo mais interessante

Agora vamos calcular os seus pontos de experiência de um modo mais interessante. 

Digamos que os pontos de experiência que você ganha dependem do seu nível de experiência e que quanto maior o seu nível mais pontos de experiência você ganha por cada robô feiticeiro capturado.

Para cada nível de experiência atingido você ganhará 1 ponto de experiência de bônus além dos pontos normalmente ganhos por cada robô capturado.

<center>

!['XP: 20 Placar: ???????'](img/xp20-placar.png)

!['robô feiticeiro 1'](img/robof-64px.png)
!['robô feiticeiro 2'](img/robof-64px.png)
!['robô feiticeiro 3'](img/robof-64px.png)
!['robô feiticeiro 4'](img/robof-64px.png)
!['mago'](img/mago-64px.png)

</center>


Digamos que seu nível de experiência é `20`. Então, se ao capturar 4 robôs você ganha `4 * 0.05` pontos de experiência, agora você vai ganhar esse valor e mais `4 * 20`, que são 20 pontos de experiência por cada robô capturado.

O cálculo fica assim:

In [6]:
# Cálculo dos pontos de magia ganhos.
# Você capturou 4 robôs feiticeiros
# e cada um deles vale 0.5 pontos
# de magia.

4 * 0.5

2.0

In [7]:
# Cálculo do bônus por nível de experiência.
# Supondo que seu nível atual é 20.
# Então, são 20 pontos por cada robô capturado.

4 * 20

80

Finalizando a conta, agora segundo essa nova forma de calcular os pontos, os seus pontos por ter capturado os 4 robôs nessa jogada serão:

In [8]:
0.2 + 2.0 + 80

82.2

<center>

!['XP: 20 Placar: 80.2'](img/xp20-placar82.2.png)

</center>

O valor da soma dos pontos normais (`0.2`) mais os pontos de magia (`2.0`) mais o bônus por experiência (`80`) é o valor de pontos ganhos na jogada, que é igual a `82.2`. 

## Sequência de execução das operações matemáticas

As operações matemáticas não são feitas todas ao mesmo tempo. Assim como nós, quando fazemos uma conta, o computador também vai fazendo uma conta de cada vez. A ordem que o computador decide seguir para ir fazendo as contas depende dos operadores aritméticos usados no cálculo. Isso é o que se chama de **precedência de operadores**.

Alguns operadores têm precedência sobre outros. Isso quer dizer que as operações efetuadas por eles vão acontecer primeiro que as operações efetuadas por outros operadores.

Por exemplo, vamos fazer a soma dos pontos toda de uma vez, só, em apenas uma linha de código.

In [9]:
# Nesta linha de código temos 
# os 3 cálculos de pontuação
# que fizemos antes, além
# da soma de cada um dos resultados
# que é o que nos dá os pontos na jogada

# Pontos normais por robô capturado: 4 * 0.05
# + Pontos de magia: 4 * 0.5
# + Pontos de bônus por experiência: 4 * 20

4 * 0.05 + 4 * 0.5 + 4 * 20

82.2

In [24]:
# E poderia ser feito em 3 linhas diferentes?

# usando uma variável auxiliar para fazer a operação
primeira_multiplicacao = 4 * 0.05
segunda_multiplicacao = 4 * 0.5
terceira_multiplicacao = 4 * 20

x + y + z


82.2

Tivemos o mesmo resultado porque as operações foram executadas na mesma ordem que nós usamos antes, apesar de estarem todas em uma mesma linha. 

Isso aconteceu por causa da precedência que a operação de multiplicação tem sobre a operação de adição. O que isso quer dizer é que as operações de multiplicação são feitas antes das operações de adição. 

Vamos escrever novamente essa linha mas agora colocando entre parênteses as operações que são executadas primeiro. Assim vai ficar mais clara qual é a ordem de execução delas.

In [ ]:
(4 * 0.05) + (4 * 0.5) + (4 * 20)

As operações que estão entre parênteses são executadas primeiro. Elas são executadas da esquerda para a direita.

Depois que cada operação de multiplicação é executada, o resultado de cada uma é colocado no lugar de cada operação. Isso acontece na memória do computador, mas se fossemos escrever em código, ficaria assim:


In [ ]:
0.2 + 2.0 + 80

Essas somas agora são executadas, uma por vez, da esquerda para a direita. O resultado é o mesmo de antes: `82.2`.

## Alterando a precedência das operações

Podemos usar os parênteses para alterar a sequência que as operações são executadas pelo Python. 

Por exemplo:

In [27]:
4 * (0.05 + 4) * 0.5 + 4 * 20

88.1

Essa conta já não tem nada a ver com o cálculo do score do mago, mas vamos observar o resultado para aprender sobre precedência de operadores.

O operador `+` não tem precedência sobre o `*`, mas como colocamos a operação de adição entre parênteses, ela foi executa primeiro mesmo assim. Só depois é que a multiplicação por `4` foi feita. Isso alterou o resultado para `88.1`, porque o cálculo que fizemos agora foi de fato diferente do anterior. 

Nesse caso, o que dissemos foi:

> "some 0.05 com 4, depois multiplique o resultado por 4 e depois multiplique esse novo resultado por 0.5 e depois multiplique 4 por 20 e agora some esses dois resultados".



A sequência é esta:

In [28]:
0.05+4

4.05

In [29]:
4.05 * 4

16.2

In [30]:
16.2 * 0.5

8.1

In [31]:
4 * 20

80

In [32]:
8.1 + 80

88.1

Também podemos *aninhar* parênteses. Aninhar significa colocar um parêntesis por fora de outro (ou por dentro, dependendo de como você vê). Quando aninhamos parênteses as operações que estão nos parênteses mais internos são executadas primeiro. Assim:

In [33]:
(4 * (0.05 + 4) * 0.5 + 4) * 20

242.0

Note no exemplo acima como a nova ordem de execução das operações alterou novamente o resultado final.

## Usando Variáveis

Variáveis são uma forma de darmos nomes a valores que usamos em nossos programas para guardá-los e usá-los depois em nosso código.

Para criar uma variável, escrevemos um nome para a variável e atribuímos um valor a ela usando o operador de atribuição do Python, que é o sinal `=`.

In [ ]:
velocidade = 500

O código acima cria a variável chamada `velocidade` e atribui a ela o valor `500`.

### Variáveis no lugar de valores

Quando temos uma variável podemos usar a variável em substituição ao valor que foi atribuído a ela. 

No caso do cálculo da pontuação do mago em uma jogada, podemos escrever uma variável para cada um dos valores usados para calcular os pontos e, depois, reescrever a fórmula que calcula a pontuação usando apenas as variáveis no lugar dos valores.

Vamos ver como fazer isso criando uma variável de cada vez.

   
!['xp: 20'](img/xp20.png)
    
Começamos pelo nível de experiência do jogador, pode ser guardado em uma variável.

In [6]:
experiencia = 20

Lemos essas linhas de código assim:

- experiencia é igual a 20

ou

- experiencia contém 20

ou, ainda

- experiencia é 20


A forma de falar o que está escrito no código vai variar conforme a situação. Algumas vezes vai fazer mais sentido dizer *é igual a* e outras vai fazer mais sentido dizer *contém* ou *é*...

> Note que o nome da variável que usamos para escrever o código do programa não precisa ser igual ao nome que vamos mostrar para quem for usá-lo. Neste caso a variável chama `experiencia` mas no placar estamos chamando de `xp`. Quando chegarmos na programação de interfaces gráficas vamos ver melhor como a apresentação dos dados é independente do modo como os dados são armazenados na memória do computador.

<center>

!['pts: 0.05 | magia: 0.5'](img/pts-005-magia-05.png)

</center>

Guardamos os pontos de experiência ganhos por capturar um robô e também os pontos de magia. Criamos uma variável para cada um.

In [4]:
pts_experiencia = 0.05
pts_magia = 0.6

Precisamos guardar o número de robôs que capturamos na jogada. Esse valor não só faz parte do cálculo da pontuação como também é um dos que mais vai variar durante um jogo, porque a cada jogada o número de robôs capturados pode mudar. Ele pode ser 4, 2, zero, 1.000...

<center>

!['robô feiticeiro 1'](img/robof-64px.png)
!['robô feiticeiro 2'](img/robof-64px.png)
!['robô feiticeiro 3'](img/robof-64px.png)
!['robô feiticeiro 4'](img/robof-64px.png)

</center>

In [2]:
robos = 5

Agora, podemos reescrever o cálculo dos pontos ganhos por capturar um robô usando as variáveis no lugar onde escreveríamos esses valores.

In [7]:
(robos * pts_experiencia) + (robos * pts_magia) + (robos * experiencia)

103.25

### Variáveis como rótulos

No exemplo anterior pudemos ver como a a fórmula que calcula a pontuação ficou mais clara quando escrevemos ela usando variáveis, já que agora não vemos apenas números na linha de código. 

> O nome de uma variável pode ser usado para explicar (se for bem escolido) o que é que a variável contém.

É como se o nome da variável fosse um rótulo que colocamos em uma caixa onde guardamos coisas. A caixa seria um espaço na memória do computador e a variável seria um rótulo que colocamos nessa caixa. Depois de colocar esse rótulo, podemos usar o rótulo para nos referirmos tanto à caixa quando ao conteúdo da caixa. 

Por exemplo, imagine uma caixa com robôs. Na caixa colamos um rótulo onde está escrito `Robôs Capturados`.

<center>

!['Robôs Capturados'](img/caixa_robos_capturados.png)

</center>

Em Python, podemos usar uma `lista` para representar essa caixa. Mais adiante vamos ver como trabalhar com listas. Por enquanto, vamos escrever essa linha de código apenas para ilustrar o que está sendo explicado.

Execute o código abaixo para acompanhar os próximos exemplos.

In [13]:
Robos_capturados = ['robô 1', 'robô 2', 'robô 3', 'robô 4']


Agora, se alguém nos disser "pegue os robôs capturados", podemos entender isso de duas formas: 

1. pegar a caixa 'Robôs Capturados'
1. pegar o conteúdo da caixa 'Robôs Capturados'

Se pegamos a caixa, levamos junto seu conteúdo. Se pegamos só o conteúdo, ora, então pegamos o conteúdo. Em qualquer caso, acabamos sempre pegando o conteúdo da caixa que, nesse caso, é o valor da variável. O que acontece é que às vezes podemos também pegar a caixa junto e acabamos com duas coisas em mãos: os robôs (conteúdo) e a caixa (a variável).

Por exemplo, execute os dois trechos de código abaixo e veja como o efeito é o mesmo.

In [9]:
# Assim estamos pegando "só o conteúdo da caixa"

['robô 1', 'robô 2', 'robô 3', 'robô 4']

['robô 1', 'robô 2', 'robô 3', 'robô 4']

In [10]:
# Assim estamos pegando "a caixa e seu conteúdo"

robos_capturados

['robô 1', 'robô 2', 'robô 3', 'robô 4']

In [15]:
robos_capturados[1]

'robô 2'

In [16]:
robos_capturados

['robô 1', 'robô 2', 'robô 3', 'robô 4']

In [17]:
botos_capturados = ['robô 1', 'robô 2', 'robô 3', 'robô 4']

In [18]:
botos_capturados

['robô 1', 'robô 2', 'robô 3', 'robô 4']

In [1]:
robos_capturados = ['robô 1', 'robô 2', 'robô 3', 'robô 4']
robos_capturados = ['robô 1', 'robô 2']
print(robos_capturados)

['robô 1', 'robô 2']



No caso da linha de código que escrevemos antes, quando escrevemos o nome de uma variável estamos pegando a caixa e também seu conteúdo. Quando o Python está fazendo os cálculos, ele está usando o nome que demos à caixa (o nome da variável) para pegar o conteúdo que está na caixa (na memória do computador) e fazer o cálculo usando só o conteúdo (o valor da variável).

Por isso também dizemos que uma variável *guarda* ou *armazena* um valor. Na verdade, é o computador que guarda o valor em sua memória, mas como através da variável podemos executar essa operação de armazenamento e recuperação de dados, fica prático para nós, que programamos, pensar na própria variável como meio e local de armazenamento de dados.


### Variáveis para acessar e modificar um valor na memória

A segunda (e grande) vantagem é que também podemos usar o nome da variável para alterar o valor que está guardado na memória do computador. 

Por exemplo, se agora alterarmos o valor da variável `robos` para `2` (porque nessa jogada capturamos dois robôs e não 4 como da outra vez), então a mesma fórmula vai dar um resultado diferente.

<center>

!['XP: 20 Placar: 41.1'](img/xp20-placar41.1.png)

</center>

In [ ]:
robos = 2
(robos * pts_experiencia) + (robos * pts_magia) + (robos * experiencia)

Mudamos o valor da variável `robos` para `2` e agora o resultado é `41.1`.

Se mudamos o valor de `robos` para `5`, qual o resultado? Experimente alterando o valor da variável `robos` na célula acima e executando novamente o código.

## Fluxo de execução de um programa

Note que no último exemplo temos duas linhas de código. Elas são executadas em sequência, de cima para baixo.

A ordem da execução de cada linha de código é muito importante. Como queremos calcular a pontuação usando a mesma fórmula mas com um número de robôs diferente, precisamos *primeiro* mudar o número de robôs para *depois* calcular a pontuação. Se não fizermos assim, o cálculo da pontuação vai ser feito com um número de robôs desatualizado e vai sair errado.

O código de um programa é executado sempre de cima para baixo. Chama-se isso de `fluxo de execução`. Porém, às vezes, podem haver desvios do fluxo de execução. Isso acontece quando redirecionamos o fluxo de execução para outro código, que é uma outra parte do programa. 

Os desvios de execução podem acontecer basicamente em três casos:

- quando uma função é executada
- quando uma linha de código que contém uma sentença condicional é encontrada
- quando o programa executa um loop 

Vamos ver os três casos mais adiante.

## Usando uma função para mostrar o valor de uma variável

Vamos falar agora em *impressão*, mas não se trata de impressão em papel. É um pouco estranho na primeira vez que ouvimos isso, mas em programação chamamos a operação que exibe um texto ou números na tela de `imprimir`.

Também falamos `exibir` ou `mostrar`, que parecem fazer mais sentido para nós hoje, mas é que o termo `imprimir` vem de muito tempo atrás, quando a principal forma de os computadores exibirem os resultados das suas operações era através de impressoras que imprimiam eles em papel!

Em Python, para *imprimir* o valor de uma variável, usamos uma função chamada `print()`. 

> Uma função é uma operação pré-programada: um trecho de código que podemos executar a qualquer momento.

Por exemplo, imprimimos o valor da variável `experiencia` assim:

In [2]:
experiencia = 20
print(experiencia)
experiencia = 30
print(experiencia)

20
30


A função `print()` escreve na tela o valor da variável.

> Note que a variável é escrita entre os parênteses após o nome da função.

E se quiséssemos imprimir o resultado da pontuação? Uma forma de fazer isso é guardando esse resultado em uma variável para depois usar a função `print()` para imprimir o valor dessa variável.

In [ ]:
robos = 2
resultado = (robos * pts_experiencia) + (robos * pts_magia) + (robos * experiencia)

print(resultado)

A função print também pode imprimir qualquer texto. Por exemplo, alguma mensagem que você queira exibir.

In [ ]:
print('Você ganhou pontos!')

Podemos "misturar" nessa mensagem o valor da variável `resultado` e exibir uma mensagem com o número de pontos que o mago ganhou na jogada.

Para isso, precisamos aprender a trabalhar com `Strings`, que são um tipo de dado diferente dos números com os quais trabalhamos até agora. Vamos ver isso a seguir.

Por agora, execute a linha de código abaixo para ver como fica a mensagem completa.

In [ ]:
print(f'Você ganhou {resultado} pontos!')

## Juntando tudo para nosso primeiro programa!

O código abaixo é o resultado de todas as linhas de código que precisamos para calcular o resultado de uma jogada. Esse é nosso primeiro programa!

In [2]:
experiencia = 20
pts_experiencia = 0.05
pts_magia = 0.5


robos = int(input('Oh, honorável Mago! Quantos robôs capturaste? '))

resultado = (robos * pts_experiencia) + (robos * pts_magia) + (robos * experiencia)

print(f'Você ganhou {resultado} pontos!')
    

Oh, honorável Mago! Quantos robôs capturaste?  4


Você ganhou 82.2 pontos!


Esse código pode ser salvo em um arquivo e executado em seu computador. Para isso você vai precisar de um editor de texto como o `Bloco de Notas` do Windows ou o `Notepad++` ou qualquer outro editor que trabalhe com "texto simples". 


Um arquivo de texto simples costuma ter a extensão .TXT e é diferente de um arquivo de texto como um .DOCX do Word ou um PDF. Qualquer programa de computador é escrito e salvo em um arquivo de texto simples. Esse é um texto sem formatações especiais (como fontes, cores, imagens, etc...) e contém apenas caracteres. Caracteres são qualquer símbolo que você pode digitar no teclado, inclusive números e outros caracteres especiais, como `$`, `*`, `@`, `%`, etc...

Supondo que você esteja usando o Windows, abra o `Bloco de Notas` e copie o código da célula acima em um novo arquivo e salve ele como 'pontos_na_jogada.py'. Depois, abra o terminal do Windows, vá para o diretório onde você salvou o arquivo e execute ele com Python com a seguinte linha de comando:

<pre>

python pontos_na_jogada.py

</pre>

---

<center>

!['Fim da parte 01'](img/fim_parte_01.png)

!['Continuar (S/n)?'](img/Continuar-Sn.png)

</center>

In [2]:
experiencia = 20
pts_experiencia = 0.05
pts_magia = 0.5

continuar = 's'

while continuar == 's':
    robos = int(input('Oh, honorável Mago! Quantos robôs capturaste? '))

    resultado = (robos * pts_experiencia) + (robos * pts_magia) + (robos * experiencia)

    print(f'Você ganhou {resultado} pontos!')
    continuar = input('Continuar (s/n)?' )

print('Fim do programa')

Oh, honorável Mago! Quantos robôs capturaste?  5


Você ganhou 102.75 pontos!


Continuar (s/n)? s
Oh, honorável Mago! Quantos robôs capturaste?  2


Você ganhou 41.1 pontos!


Continuar (s/n)? n


Fim do programa
